File: 01_strings_and_arrays/06_pair_product/pair_product.py
def pair_product(numbers, target_product):
  map = {}

  for index, num in enumerate(numbers):
    target_value = target_product / num

    if target_value in map:
      return (index, map[target_value])
    else:
      map[num] = index

print(pair_product([3, 2, 5, 4, 1], 8)); # -> [1, 3]
print(pair_product([3, 2, 5, 4, 1], 10)); # -> [1, 2]
print(pair_product([4, 7, 9, 2, 5, 1], 5)); # -> [4, 5]
print(pair_product([4, 7, 9, 2, 5, 1], 35)); # -> [1, 4]
print(pair_product([3, 2, 5, 4, 1], 10)); # -> [1, 2]
print(pair_product([4, 6, 8, 2], 16)); # -> [2, 3]

File: 01_strings_and_arrays/06_pair_product/pair_product.js
const pairProduct = (numbers, targetProduct) => {
    // Approach - Track indicies with hash map, compare current element to target value
    // 1. Establish a hash map
    // 2. Loop over array
    // 2*. Establish a target value (targetProduct / ele = targetEle) if prduct % ele = 0
    // 2*. If match is found, return the index from map and current ele index
    // 2*. Else add ele to hash map
    const map = {}

    for(idx in numbers){
        const targetValue = targetProduct / numbers[idx]

        if(targetValue in map) {
            return [Number(idx), Number(map[targetValue])]
        } else {
            map[numbers[idx]] = idx 
        }
    }

  };
  
console.log(pairProduct([3, 2, 5, 4, 1], 8)); // -> [1, 3]
console.log(pairProduct([3, 2, 5, 4, 1], 10)); // -> [1, 2]
console.log(pairProduct([4, 7, 9, 2, 5, 1], 5)); // -> [4, 5]
console.log(pairProduct([4, 7, 9, 2, 5, 1], 35)); // -> [1, 4]
console.log(pairProduct([3, 2, 5, 4, 1], 10)); // -> [1, 2]
console.log(pairProduct([4, 6, 8, 2], 16)); // -> [2, 3]

  module.exports = {
    pairProduct,
  };
  